### PR TITLE
Fixed number formatting and added tests

### LINKED ISSUE #3080: Consistent handling and formatting of numbers
The formatting of numbers differs in various places. Just take a look at the "money earned" view under Stats.

image

The standard ns.nFormat caps out at "t".

The formatting should be the same everywhere (and support larger values), like in most of the corporation stuff. One issue there though seems to be that you can't input in the same format. For example, under factions can you choose to donate "1t" but that doesn't work for the corp inputs.

One suggestion regarding this from HeinousTugboat was to take inspiration from [Tyrope's lib](https://github.com/tyrope/bitburner/blob/master/lib/format.js])

Bitburner v1.4.0

### PR DESCRIPTION
**Pull request by Master-Guy**
Added test file: test\jest\ui\nFormat.test.js
image

List of changes:
- Correctly parse negative numbers with extraNotations
- Correctly parse really big numeric strings with extraNotations
- Parse really big negative numeric strings with extraNotations
-- (solves not-yet-reported bug: '-123.456k' would result in -123 instead of -123456)
- Decimals in formatBigNumber are now optional
- Decimals in formatReallyBigNumber are now optional

Closes #3080

### CODE REVIEW

**Comment by phyzical** [comment]
tests looks to cover things fairly well, but mind a quick gif  of it in places of the game itself

File: src/ui/numeralFormat.ts [hunk 1]
```diff
@@ -187,19 +188,52 @@ class NumeralFormatter {
...
+    }
+    const [, number, notation] = match;
+    const notationIndex = extraNotations.indexOf(notation);
+    if (notationIndex === -1) {
+      return NaN;
+    }
+    return parseFloat(number) * extraFormats[notationIndex];
+  }
+
+  furthestFrom0(n1: number, n2 = 0, n3 = 0): number {
```
[the remark this answers is not in the document]
Comment (Master-Guy) [reply src/ui/numeralFormat.ts:205]: As we are parsing the same number 3 times, there can't be an instance where one is positive and the other is negative.
Even if there was, it would return the lowest number, which would be accurate, as we are working with a negative number in that case.

[the remark this answers is not in the document]
Comment (ApamNapat) [reply src/ui/numeralFormat.ts:205]: What I meant is - this function doesn't check the input in any way and if you pass different numbers, then you can get results that are different from what the function's name says. So maybe a docstring or a different name for the function would be in order?

[the remark this answers is not in the document]
Comment (Master-Guy) [reply src/ui/numeralFormat.ts:205]: This should be better then?

[the remark this answers is not in the document]
Comment (Master-Guy) [reply src/ui/numeralFormat.ts:205]: https://github.com/danielyxie/bitburner/pull/3150/commits/1d07bf049ff9957831ceac21f5e5692aad86031c

[the remark this answers is not in the document]
Comment (phyzical) [reply src/ui/numeralFormat.ts:205]: how about instead of `furthestFrom0` ``largestAbsoluteNumber`

[the remark this answers is not in the document]
Comment (Master-Guy) [reply src/ui/numeralFormat.ts]: Renamed and pushed

File: test/jest/ui/nFormat.test.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (ApamNapat) [inline test/jest/ui/nFormat.test.js:1]: Why is this needed?

Comment (Master-Guy) [reply test/jest/ui/nFormat.test.js:1]: Because of the unused import of `jest`. (Copied from DynamicRamCalculation.test.js, not sure if I can remove it without consequences)

**Review by phyzical — APPROVED** [review]
